File: Loon-0.5-Alpha/Loon-Neo/src/loon/LGame.java
/**
 * Copyright 2008 - 2015 The Loon Game Engine Authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.5
 */
package loon;

import loon.event.InputMake;
import loon.utils.reply.Act;

/*
 * 这类为最主要的游戏功能类集合对象，其中涵盖了Loon的基础对象实例。
 */
public abstract class LGame {

	public Act<Error> errors = Act.create();

	public LGame(final LSetting config) {
		if (config == null) {
			this.setting = new LSetting();
		}
		this.setting = config;
		String appName = config.appName;
		if (appName != null) {
			LSystem.APP_NAME = appName;
		} else if (LSystem.APP_NAME != null) {
			appName = LSystem.APP_NAME;
		} else {
			appName = "loon";
			LSystem.APP_NAME = appName;
		}
		setting.appName = appName;
		String fontName = config.fontName;
		if (fontName != null) {
			LSystem.FONT_NAME = fontName;
		} else if (LSystem.FONT_NAME != null) {
			fontName = LSystem.FONT_NAME;
		} else {
			fontName = "Dialog";
			LSystem.FONT_NAME = fontName;
		}
		setting.fontName = fontName;
	}

	public void initProcess() {
		LSystem.init(this);
	}

	public static enum Type {
		JAVASE, ANDROID, IOS, WP, HTML5, STUB
	}

	public LSetting setting;

	private Display display;

	public static enum Status {
		PAUSE, RESUME, EXIT
	};

	public Act<Status> status = Act.create();

	public Act<LGame> frame = Act.create();

	public static class Error {
		public final String message;
		public final Throwable cause;

		public Error(String message, Throwable cause) {
			this.message = message;
			this.cause = cause;
		}
	}

	private static Class<?> getType(Object o) {
		if (o instanceof Integer) {
			return Integer.TYPE;
		} else if (o instanceof Float) {
			return Float.TYPE;
		} else if (o instanceof Double) {
			return Double.TYPE;
		} else if (o instanceof Long) {
			return Long.TYPE;
		} else if (o instanceof Short) {
			return Short.TYPE;
		} else if (o instanceof Short) {
			return Short.TYPE;
		} else if (o instanceof Boolean) {
			return Boolean.TYPE;
		} else {
			return o.getClass();
		}
	}

	public Display register(Class<? extends Screen> clazz, Object... args) {
		LSystem.viewSize.setSize(setting.width, setting.height);
		this.display = new Display(this, setting.fps);
		if (clazz != null) {
			if (args != null) {
				try {
					final int funs = args.length;
					if (funs == 0) {
						display.setScreen(clazz.newInstance());
					} else {
						Class<?>[] functions = new Class<?>[funs];
						for (int i = 0; i < funs; i++) {
							functions[i] = getType(args[i]);
						}
						java.lang.reflect.Constructor<?> constructor = Class
								.forName(clazz.getName()).getConstructor(
										functions);
						Object o = constructor.newInstance(args);
						if (o != null && (o instanceof Screen)) {
							display.setScreen((Screen) o);
						}
					}
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		}
		return display;
	}

	public void reportError(String message, Throwable cause) {
		errors.emit(new Error(message, cause));
		log().warn(message, cause);
	}

	public <E> void dispatchEvent(Act<E> signal, E event) {
		try {
			signal.emit(event);
		} catch (Throwable cause) {
			reportError("Event dispatch failure", cause);
		}
	}

	protected void emitFrame() {
		try {
			frame.emit(this);
		} catch (Exception e) {
			log().warn("Frame tick exception", e);
		}
	}

	public void invokeLater(Runnable runnable) {
		asyn().invokeLater(runnable);
	}

	public boolean isAsyncSupported() {
		return asyn().isAsyncSupported();
	}

	public void invokeAsync(Runnable action) {
		asyn().invokeAsync(action);
	}

	public abstract LGame.Type type();

	public abstract double time();

	public abstract int tick();

	public abstract void openURL(String url);

	public abstract Assets assets();

	public abstract Asyn asyn();

	public abstract Graphics graphics();

	public abstract InputMake input();

	public abstract Json json();

	public abstract Log log();

	public abstract Save save();

	public abstract Support support();

	public final Display display() {
		return display;
	}

}

File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/LGame.cs
﻿using java.lang;
using loon.events;
using loon.utils;
using loon.utils.reply;
using Microsoft.Xna.Framework.Graphics;
using static loon.LTexture;

namespace loon
{
    public abstract class LGame : LRelease
    {
        public enum Type
        {
            JAVASE, MONO, ANDROID, IOS, WP, HTML5, UNITY, SWITCH, PS, XBOX, STUB
        }

        public enum Status
        {
            PAUSE, RESUME, EXIT
        }

        public class Error
        {

            public string message;
            public System.Exception cause;

            public Error(string message, System.Exception cause)
            {
                this.message = message;
                this.cause = cause;
            }
        }


        protected internal const string FONT_NAME = "Dialog";

        protected internal const string APP_NAME = "Loon";

        protected internal static LGame _base = null;

        protected internal static Platform _platform = null;

        // 错误接口
        public Act<Error> errors = Act<Error>.Create<Error>();

        // 状态接口
        public Act<Status> status = Act<Status>.Create<Status>();

        // 游戏窗体刷新接口
        public Act<object> frame = Act<object>.Create<object>();

        // 游戏基本设置
        public LSetting setting;

        protected internal LProcess processImpl;

        protected internal Display displayImpl;

        public LGame(LSetting config, Platform plat)
        {
            LGame._platform = plat;
            if (config == null)
            {
                config = new LSetting();
            }
            this.setting = config;
            string appName = config.appName;
            if (StringUtils.IsEmpty(appName))
            {
                setting.appName = APP_NAME;
            }
            string fontName = config.fontName;
            if (StringUtils.IsEmpty(fontName))
            {
                setting.fontName = FONT_NAME;
            }

        }

        protected void InitProcess(LGame game)
        {
            _base = game;
            if (_base == null && _platform != null)
            {
                _base = _platform.GetGame();
            }
            processImpl = new LProcess(game);
            Log().Debug("The Loon Game Engine is Begin");
        }

        public LGame InitProcess()
        {
            InitProcess(this);
           /* if (setting.defaultGameFont == null)
            {
                setting.defaultGameFont = LFont.getFont(setting.fontName, 20);
            }
            if (setting.defaultLogFont == null)
            {
                setting.defaultLogFont = LSTRFont.getFont(LSystem.isDesktop() ? 16 : 20);
            }
            if (jsonImpl == null)
            {
                jsonImpl = new JsonImpl();
            }*/
            return this;
        }

        public virtual LTexture CreateTexture(int width, int height, Format config)
        {
            return Graphics().CreateTexture(width, height, config);
        }

        public virtual LTexture NewTexture(string path)
        {
            return NewTexture(path, Format.LINEAR);
        }


        public virtual LTexture NewTexture(string path, Format config)
        {
            if (StringUtils.IsEmpty(path))
            {
                return null;
            }
            Log().Debug("Texture : New " + path + " Loaded");
            return this.Graphics().CreateTexture(path,config);
        }

        public virtual void SetPlatform(Platform plat)
        {
            if (plat != null)
            {
                LGame._platform = plat;
                LGame game = plat.GetGame();
                if (game != null)
                {
                    LGame._base = game;
                  //  LGame._base.resetShader();
                }
            }
        }
        protected internal virtual LGame CheckBaseGame(LGame game)
        {
            LGame oldGame = _base;
            if (game != oldGame && game != null)
            {
                oldGame = game;
            }
            else if (game == null)
            {
                if (oldGame != null && _platform != null && _platform.GetGame() != null)
                {
                    if (oldGame != _platform.GetGame())
                    {
                        oldGame = _platform.GetGame();
                    }
                }
            }
            if (_base != game || _base != oldGame)
            {
                _base = oldGame;
            }
            return LSystem.Base;
        }

        public virtual LProcess Process()
        {
            return this.processImpl;
        }

        public virtual LGame ReportError(string message, System.Exception cause)
        {
            return ReportError(message, cause, true);
        }


        public virtual LGame ReportError(string message, System.Exception cause, bool logError)
        {
            errors.Emit(new Error(message, cause));
            if (logError)
            {
                Log().Error(message, cause);
            }
            return this;
        }

        public virtual void DispatchEvent<E>(Act<E> signal, E e)
        {
            try
            {
                signal.Emit(e);
            }
            catch (System.Exception cause)
            {
                ReportError("Event dispatch failure", cause);
            }
        }

        protected internal virtual void EmitFrame()
        {
            try
            {
                frame.Emit(this);
            }
            catch (System.Exception cause)
            {
                Log().Error("Frame tick exception :", cause);
                LSystem.StopRepaint();
            }
        }

        public virtual Display Register(Screen screen)
        {
            this.displayImpl = new Display(this, setting.fps);
            this.displayImpl.SetScreen(screen);
            return displayImpl;
        }

        public static void FreeStatic()
        {
            LGame._platform = null;
            LGame._base = null;
        }

        public virtual bool IsMobile()
        {
            Type type = this.TYPE;
            return (type == LGame.Type.ANDROID || type == LGame.Type.IOS || type == LGame.Type.WP
                    || type == LGame.Type.SWITCH);
        }

        public virtual bool IsHTML5()
        {
            Type type = this.TYPE;
            return type == LGame.Type.HTML5;
        }

        public virtual bool IsDesktop()
        {
            return !IsMobile() && !IsHTML5();
        }

        public abstract LGame.Type TYPE { get; }
        public abstract double Time();
        public abstract int Tick();

        public virtual Display Display()
        {
            return displayImpl;
        }
        public abstract Log Log();
        public abstract Asyn Asyn();
        public abstract Assets Assets();
        public abstract Support Support();
        public abstract InputMake Input();

        public abstract Graphics Graphics();

        public virtual void Close()
        {
            if (!errors.IsClosed())
            {
                errors.ClearConnections();
            }
            if (!status.IsClosed())
            {
                status.ClearConnections();
            }
            if (!frame.IsClosed())
            {
                frame.ClearConnections();
            }
        }

    }
}
